### ISSUE TITLE
`COLLATION` Support

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#9192]
### Is your feature request related to a problem or challenge?

"Collation" generically means how to compare and sort string values. 

Databases like Postgres allow you to change the default collation order to control this more carefully to match whatever the user wants rather than what the standard sort order means

Someone asked about this on discord: https://discord.com/channels/885562378132000778/1166447479609376850/1205554368292855868

Here are some details on how this works in Postgres: https://www.postgresql.org/docs/current/collation.html

### Describe the solution you'd like

Someone to design and implement `COLLATION`

This probably looks like a `SessionConfig` setting to control collation at the session level and possibly some way to define it as part of the table definition

**Comment by alamb** [comment]
cc @gruuya  as I believe you mentioned you might also be interested in this feature

**Comment by tustvold** [comment]
I think we would need to identify a mature Rust library for supporting collations, as I suspect this is not something we wish to implement ourselves

**Comment by alamb** [comment]
I believe `collation` is an important feature for postgres compatibility, but it is not as widely used in other databases. 

I agree having something that was optional would be ideal

**Comment by gruuya** [comment]
Yeah, for reference DuckDB also provides optional general collations, using an extension for the ICU project: https://duckdb.org/docs/sql/expressions/collations.html#icu-collations 

There's a corresponding Rust crate as well: https://docs.rs/icu/latest/icu/collator/index.html

**Comment by andygrove** [comment]
A major feature of Spark 4.0.0 is support for string collation

**Comment by mbutrovich** [comment]
> There's a corresponding Rust crate as well: https://docs.rs/icu/latest/icu/collator/index.html

There's also Google's [rust_icu](https://github.com/google/rust_icu).

**Comment by alamb** [comment]
I think it would be great to sketch out what supporting collation would look like.

As I recall, it is fairly invasive, as it potentially changes all string comparisons (both for ORDER BY as well as for `=`, `<`, etc type functions).

The first major question, is where the support would live (upstream in arrow-rs or in DataFusion) 🤔

**Comment by mjgarton** [comment]
It seems collation support is being discussed for the arrow c++ implementation.  I don't know if this implies arrow-rs is a more suitable place than datafusion?

https://github.com/apache/arrow/issues/27875

Datafusion would need some degree of support in any case I suppose, even if it delegates the actual ordering decisions to arrow-rs.

**Comment by alamb** [comment]
I started a discussion on an API to support customizing behavior like this on 
- https://github.com/apache/datafusion/issues/18223

I think COLLATION would be a natural additional property for types (as it controlls comparisons and sorting)

**Comment by mjgarton** [comment]
I notice that support for extension types has been merged.  I can't quite see how to get from that to collation support.  Is more work needed in extension types first?
